tgl: enable WOVCRO clock

abonislawski opened the pull request:
First patch will introduce CPU_WAITI_FREQ_IDX so a specific clock for WAITI can be configured per platform.
Second patch added releasing clock for waiti switches, it will fix #3448
Last patch will enable WOVCRO for TGL and allow to save significant amount of SOC power in low power S0ix WoV.

Patch is backward-compatible, if WOVCRO is not available (some old TGL steppings) then LPRO will be used for WAITI

abonislawski commented:
SOFCI TEST

keyonjie commented:
@abonislawski Thanks for enabling WOVCRO. 

My concern is that switching to WOVCRO everywhere we intent to switch to LPRO looks too aggressive to me, can we force this for WoV on S0Ix only, and make the WFI context low power clock configurable?  e.g. we already have USE_LPRO_IN_WAITI, maybe we can add a new item (e.g. USE_WOVCRO_IN_WAITI) under it and unselect it by default?

lgirdwood commented:
SOFCI TEST

abonislawski commented:
@keyonjie its not everywhere, its in WAITI only.
low_power_mode is still switching to LPRO

but we can limit it to S0ix only if needed, any pros/cons?

Anyway WOVCRO needs to be enabled by default for our releases

keyonjie commented on src/platform/intel/cavs/platform.c:
```diff
@@ -440,6 +440,16 @@ int platform_init(struct sof *sof)
+	if (shim_read(SHIM_CLKCTL) & SHIM_CLKCTL_WOV_CRO_REQUEST)
+		shim_write(SHIM_CLKCTL, shim_read(SHIM_CLKCTL) & ~SHIM_CLKCTL_WOV_CRO_REQUEST);
+	else
+		panic(SOF_IPC_PANIC_PLATFORM);
```
do wen really need panic here? can we just not use WOVCRO if it is not available?

abonislawski replied:
Yes, we can, that was just our early plan to stop using not compatible steppings. But we can still discuss this, what do you think @mmaka1  about @keyonjie idea?

lgirdwood replied:
We should actually report an error here (lower than panic but higher than normal). Ideally we should tell the driver with a new IPC so driver code can reset DSP HW etc and try and recover. i.e. any error that FW cannot recover we should send up stack so driver can attempt recovery.

abonislawski commented:
> Oh, then it's bad, how can we switch to WOVCRO then? The "pm_runtime_put(CORE_HP_CLK, cpu_get_id())" is invoked after the HPRO->WOVCRO switching? won't it switch to LPRO for WoV on S0Ix?

I need to investigate it because it looks like low_power_mode is a good place for WOVCRO switching. Power measurements gave great results so WOVCRO was working, need to check if this low_power_mode is working correctly then.

> I am suggesting adding config items, to make it configurable. Please be noticed that the CONFIG_CAVS_USE_LPRO_IN_WAITI can be unslected so you might need to change the platform_clock_on_waiti() of the "#else" version also.

The CAVS approach is to use WOVCRO as lowest clock in WAITI stage because of power savings, so I think we should rename CONFIG_CAVS_USE_LPRO_IN_WAITI to something more universal for LPRO/WOVCRO and another layer for LPRO-or-WOVCRO switching is not necessary.

zrombel commented:
This PR force us to update our CI TGL RVP platforms to B0 stepping. At this point we are using A2 stepping platforms thus TGL smoke tests will FAIL. We plan to finish platform upgrade at the end of WW40.

abonislawski commented:
Updated PR with clock switching moved to "low_power_mode".
The names are very confusing because the true clock switching for CONFIG_CAVS_USE_LPRO_IN_WAITI takes place in platform_clock_low_power_mode() instead of just platform_clock_on_waiti() as the name suggests.

Also added releasing clock for waiti switches, it will fix #3448

abonislawski commented:
@keyonjie good news for old steppings, added backward-compatible version (LPRO automatically selected) because it was reported to us that so many boards are still Ax :(

lgirdwood commented:
@abonislawski can you check CI could be a server issue ?

abonislawski commented:
> @abonislawski can you check CI could be a server issue ?

I think CI issue

keyonjie commented:
@abonislawski sorry, with this new WoVCRO clock introduced, the clock switching logic is becoming more complex and error-prone, we should be very careful when implement this, several items are not clear to me:

1. the clock switching logic, e.g. we have 3 clocks (clock 0/1/2) in total, and all clock switching are performed at _waiti()/_wakeup(), basically:
_waiti(): possible High -> Low switching, scenarios including:
a. no switching needed. e.g. only 1 clock is supported/configured for the specific platform, or no any low power mode(e.g. LP for WAITI, or LP for D0I3) is configured. Another scenario like LPRO -> LPRO so no switching needed.
b. clock 2 (HPRO) -> clock 1 (LPRO). e.g. LP for waiti or LP for D0I3(but no clock 0 /WOVCRO existed) low power mode is selected.
c. clock 2 -> clock 0 (WoVCRO). LP for D0I3 low power mode is selected and WoVCRO existed. Need to make sure we only do a single switching, neither 'clock 2 -> clock 1 -> clock 0' nor 'clock 2 -> clock 0 -> clock 1' switching should happen.

2. make it clear about what is configurable and what isn't. e.g. LP low power for waiti (CONFIG_CAVS_USE_LPRO_IN_WAITI at the moment), LP low power mode for D0Ix (extra configure item needed? suppose we will always be required to switch to low power mode for D0Ix entry if possible, maybe no needed here), WOVCRO on D0Ix (a new configure item needed, at least for debugging purpose, it can be default 'Y', bug can be selected as 'N' to try only LPRO for D0Ix WoV).

So in short let's make this more clear in one shot, to avoid possible mysterious errors happen in the future.

lgirdwood commented on src/include/sof/lib/clk.h:
```diff
@@ -38,6 +38,7 @@ struct clock_info {
 	const struct freq_table *freqs;
 	uint32_t default_freq_idx;
 	uint32_t current_freq_idx;
+	uint32_t waiti_freq_idx;
```
Ok, I'm reading the comments about things becoming more complex so I think we can simplify by adding an inline comment to each "idx" and "notification" member in this struct. This should start to make things simpler for all reviewers to follow.

mmaka1 replied:
Should this index be a common parameter? Seems to be specific for cAVS platforms and left uninitialized/unused for others.

lgirdwood commented on src/platform/intel/cavs/lib/clk.c:
```diff
@@ -166,14 +166,9 @@ static void platform_clock_low_power_mode(int clock, bool enable)
-		 * not release active clock just for waiti,
-		 * so they can be switched without delay on wake up.
-		 */
-		select_cpu_clock(waiti_freq_idx, false);
```
Can we add another comment here explaining the new logic.

abonislawski replied:
no logic changes in final version

lgirdwood replied:
ping

abonislawski replied:
This comment was only about delay which isnt important for us, the remaining logic is pretty simple just as 'if' in code, if clock > LPRO then change to LPRO

lgirdwood commented on src/platform/intel/cavs/lib/clk.c:
```diff
@@ -293,18 +293,29 @@ static int clock_platform_set_cpu_freq(int clock, int freq_idx)
 
 void platform_clock_init(struct sof *sof)
 {
+	uint32_t platform_waiti_clock = 1;
```
Probably better to set this to CPU_WAITI_FREQ_IDX here otherwise it can be mistaken for magic number.

abonislawski replied:
yeap, much better, fixed

lgirdwood commented on src/platform/intel/cavs/lib/clk.c:
```diff
@@ -293,18 +293,29 @@ static int clock_platform_set_cpu_freq(int clock, int freq_idx)
+	if (shim_read(SHIM_CLKCTL) & SHIM_CLKCTL_WOV_CRO_REQUEST)
+		shim_write(SHIM_CLKCTL, shim_read(SHIM_CLKCTL) & ~SHIM_CLKCTL_WOV_CRO_REQUEST);
+	else
+		platform_waiti_clock = 0;
```
Ditto, it should be set to CPU_LPRO_FREQ_IDX

lgirdwood commented on src/platform/intel/cavs/lib/clk.c:
```diff
@@ -293,18 +293,29 @@ static int clock_platform_set_cpu_freq(int clock, int freq_idx)
+#if CAVS_VERSION == CAVS_VERSION_2_5
+	/* Check HW version clock capabilities */
+
+	shim_write(SHIM_CLKCTL, shim_read(SHIM_CLKCTL) | SHIM_CLKCTL_WOV_CRO_REQUEST);
```
A small comment is needed explaining the flow and how it determines the clock to be used.

abonislawski replied:
added description

keyonjie commented:
@abonislawski ping for update.

abonislawski commented:
PR updated and rebased.
@keyonjie no logic changes for clock switching, in short: LPRO replaced by WOVCRO (if WOVCRO available)

lgirdwood reviewed (changes requested):
We've lost some context by removing a comment, we should really change the comment to match the new logic.

keyonjie commented on src/platform/tigerlake/lib/clk.c:
```diff
@@ -10,20 +10,24 @@
 };
 
 const uint32_t cpu_freq_status_mask[] = {
+	SHIM_CLKCTL_WOV_CRO_REQUEST,
```
better to redefine sth. like SHIM_CLKSTS_WOVROSCS to avoid confusion though it is actually same value.

abonislawski replied:
Sure, I will do that

lgirdwood commented:
Travis looks like a CI timeout unrelated to this PR.

keyonjie commented:
@abonislawski we should add a kconfig item to enable using this WOVCRO (e.g. disable it by default), then on most platforms where no WoV feature is needed, we use LPRO.

abonislawski commented:
Marked as WIP, will add and squash changes from tgl011 when done

keyonjie commented:
@abonislawski time to proceed and get this merged to master now, feel free to squash changes I made on the 011 release branch.

keyonjie commented:
@abonislawski ping.

lgirdwood commented:
@abonislawski @keyonjie any updates - I assume the CI is OK ?

abonislawski commented:
@lgirdwood I will updated this as soon as tgl-011 issues are fixed

lgirdwood commented:
@abonislawski thanks - this is v1.7 blocker now. I will rerun CI to make sure everything looks good with recent updates.

lgirdwood commented:
SOFCI TEST

abonislawski commented:
@lgirdwood  PR updated

lgirdwood reviewed (approved):
LGTM, but I'm not an expert on the WOVCRO. @keyonjie @mwasko @mmaka1 any comments ?
@abonislawski I assume this is same version used on production ?

abonislawski commented:
@lgirdwood yes, wovcro is already on tgl-010 and 011 releases, in 011 it was limited to s0ix only and this PR is updated with this change

keyonjie commented:
@abonislawski can you help to squash and add my commit https://github.com/thesofproject/sof/commit/dfdfa6b9403046276d8aa6ffb85c94b4dc0087f1?
Better to make sure it is exactly same with tgl-011-drop-stable for this part, which has been verified work on TGL chrome.

abonislawski commented:
@keyonjie it is already done with limiting wovcro to s0ix, this is only a few lines of code so you can easily compare that it will work exactly the same.
Cannot use your commit without any changes because it is ignoring compatibility check in platform_clock_init().

lgirdwood commented:
@abonislawski @keyonjie let me know when good to merge.

keyonjie commented on src/include/sof/lib/clk.h:
```diff
@@ -38,6 +38,7 @@ struct clock_info {
 	const struct freq_table *freqs;
 	uint32_t default_freq_idx;
 	uint32_t current_freq_idx;
+	uint32_t waiti_freq_idx;	/* lowest possible clock for waiti stage */
```
@abonislawski better to change the naming to e.g. lowest_freq_idx to avoid confusion, as we actually use LPRO for the waiti context.

abonislawski replied:
@keyonjie updated with lowest_freq_idx

keyonjie reviewed (commented):
@abonislawski yes, functional exactly same with 011 base, can you change the naming 'waiti_freq_idx' to avoid confusion, others looks good to me.

keyonjie reviewed (approved):
Looks good now, thanks.
